### PR TITLE
Filter usage for project

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #5860

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

tested manually:
create a simple setup with one project
enable usage
set a low aggregation time (10 minutes)
create a VPC with one tier and one VM in the project
after the first agregation:
2018 usage records in total
24 usage records for the project

### CODE REVIEW

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 1]
```diff
@@ -255,9 +222,7 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
         SearchCriteria<UsageVO> sc = _usageDao.createSearchCriteria();
 
         if (accountId != -1 && accountId != Account.ACCOUNT_ID_SYSTEM && !ignoreAccountId) {
-            // account exists and either domain on user role
-            // If not recursive and the account belongs to the user/domain admin, or the account was passed in, filter
-            if ((accountId == caller.getId() && !cmd.isRecursive()) || cmd.getAccountId() != null){
+            if ((accountId == caller.getId() && !cmd.isRecursive()) || cmd.getAccountId() != null || projectRequested){
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:225]: @DaanHoogland 
`accountId == caller.getId()` does not make sense

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:225]: It does, @weizhouapache it is a check whether the caller is the one asking for their own usage or someone else. Are you saying that it is not a valid test in this location? I.E. you are sugesting 
```suggestion
            if (!cmd.isRecursive()) || cmd.getAccountId() != null || projectRequested){
```
is that right?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 2]
```diff
@@ -397,6 +363,59 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
         return new Pair<List<? extends Usage>, Integer>(usageRecords.first(), usageRecords.second());
     }
 
+    private Long getDomainScopeForQuery(ListUsageRecordsCmd cmd, Long accountId, Long domainId, Account caller, boolean isDomainAdmin) {
+        if (isDomainAdmin) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:367]: We could invert this statement and return `domainId` to reduce the indentation.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 3]
```diff
@@ -397,6 +363,59 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+            }
+        }
+        return domainId;
+    }
+
+    @NotNull
+    private Long getAccountIdFromDomainPlusName(Long domainId, String accountName, Account caller) {
+        Long accountId;
+        Account userAccount = null;
+        if (_domainDao.isChildDomain(caller.getDomainId(), domainId)) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:387]: We could invert this statement to reduce indentation.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 4]
```diff
@@ -397,6 +363,59 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+
+    @NotNull
+    private Long GetAccountIdFromProject(Long projectId) {
+        Long accountId;
+        Project project = _projectMgr.getProject(projectId);
+        if (project == null) {
+            throw new InvalidParameterValueException("Unable to find project by id " + projectId);
+        }
+        final long projectAccountId = project.getProjectAccountId();
+        if(s_logger.isInfoEnabled()) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:412]:
```suggestion
        if (s_logger.isInfoEnabled()) {
```

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:367]: @GutoVeronezi I like this better
the choice is
- two indentation levels and one return/exit point vs
- one indentation and two returns/exit points
I rather have the first.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
test_nic error is not related to usage in any way I can see, retrying
@blueorangutan test

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 5]
```diff
@@ -397,6 +363,56 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+            userAccount = accounts.get(0);
+        }
+        if (userAccount == null) {
+            throw new InvalidParameterValueException("Unable to find account " + accountName + " in domain " + domainId);
+        }
+        return userAccount.getId();
+    }
+
+    @NotNull
+    private Long GetAccountIdFromProject(Long projectId) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:402]:
```suggestion
    private Long getAccountIdFromProject(Long projectId) {
```

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 6]
```diff
@@ -166,46 +168,27 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
-                }
-                if (userAccount != null) {
-                    accountId = userAccount.getId();
-                } else {
-                    throw new InvalidParameterValueException("Unable to find account " + accountName + " in domain " + domainId);
-                }
-            } else {
-                throw new PermissionDeniedException("Invalid Domain Id or Account");
-            }
+            accountId = GetAccountIdFromProject(projectId);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:181]:
```suggestion
            accountId = getAccountIdFromProject(projectId);
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Review by borisstoyanov — APPROVED** [review]
LGTM

**Review by rohityadavcloud — APPROVED** [review]

**Pull request merged**